include/directx/d3dx12.h

```diff
@@ -4037,6 +4038,432 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+            m_dOptions1.WaveLaneCountMin = 0;
+            m_dOptions1.TotalLaneCount = 0;
+            m_dOptions1.ExpandedComputeResourceStates = 0;
+            m_dOptions1.Int64ShaderOps = 0;
+        }
+
+        if (INITIALIZE_FAILED(D3D12_FEATURE_D3D12_OPTIONS2, m_dOptions2)) {
+            m_dOptions2.DepthBoundsTestSupported = false;
+            m_dOptions2.ProgrammableSamplePositionsTier = D3D12_PROGRAMMABLE_SAMPLE_POSITIONS_TIER_NOT_SUPPORTED;
+        }
+
+        if (INITIALIZE_FAILED(D3D12_FEATURE_SHADER_CACHE, m_dShaderCache)) {
+            m_dShaderCache.SupportFlags = D3D12_SHADER_CACHE_SUPPORT_NONE;
+        }
+
+        // Initialize features that requires highest version check
+        if (FAILED(m_hStatus = QueryHighestShaderModel())) {
+            return m_hStatus;
+        }
+
+        if (FAILED(m_hStatus = QueryHighestRootSignatureVersion())) {
+            return m_hStatus;
+        }
+
+        // Initialize per-node feature support data structures
+        const UINT uNodeCount = m_pDevice->GetNodeCount();
+        m_dProtectedResourceSessionSupport.resize(uNodeCount);
+        m_dArchitecture.resize(uNodeCount);
+        m_dArchitecture1.resize(uNodeCount);
+        for (UINT i = 0; i < uNodeCount; i++) {
+            m_dProtectedResourceSessionSupport[i].NodeIndex = i;
+            if (INITIALIZE_FAILED(D3D12_FEATURE_PROTECTED_RESOURCE_SESSION_SUPPORT, m_dProtectedResourceSessionSupport[i])) {
+                m_dProtectedResourceSessionSupport[i].Support = D3D12_PROTECTED_RESOURCE_SESSION_SUPPORT_FLAG_NONE;
+            }
+
+            m_dArchitecture[i].NodeIndex = i;
+            if (INITIALIZE_FAILED(D3D12_FEATURE_ARCHITECTURE, m_dArchitecture[i])) {
+                m_dArchitecture[i].TileBasedRenderer = false;
+                m_dArchitecture[i].UMA = false;
+                m_dArchitecture[i].CacheCoherentUMA = false;
+            }
+
+            m_dArchitecture1[i].NodeIndex = i;
+            if (INITIALIZE_FAILED(D3D12_FEATURE_ARCHITECTURE1, m_dArchitecture1[i])) {
+                m_dArchitecture1[i].TileBasedRenderer = false;
+                m_dArchitecture1[i].UMA = false;
+                m_dArchitecture1[i].CacheCoherentUMA = false;
+                m_dArchitecture1[i].IsolatedMMU = false;
+            }
+        }
+
+        // Initialize Feature Levels data
+        if (FAILED(m_hStatus = QueryHighestFeatureLevel())) {
+            return m_hStatus;
+        }
+        
+        return m_hStatus;
+    }
+
+    HRESULT GetStatus() const
+    {
+        return m_hStatus;
+    }
+
+    static std::variant<CD3DX12FeatureSupport, HRESULT> Create(ID3D12Device* pDevice)
+    {
+        CD3DX12FeatureSupport features;
+        HRESULT result = features.Init(pDevice);
+        if (FAILED(result)) {
+            return result;
+        }
+
+        return features;
+    }
+
+    // 0: D3D12_OPTIONS
+    FEATURE_SUPPORT_GET(BOOL, m_dOptions, DoublePrecisionFloatShaderOps);

```

Comment:
These macros are great for the implementation.  But they make it a bit harder to look at the class and see what methods can be called on it.  Perhaps all public members of the class could be predeclared at the top for the benefit of folks who don't wan't to bother sifting past implementation details?

Reply:
For the Macros, I mainly used them to simplify implementation so if there's a change in the call pattern I don't need to go and change each line but instead I can just change the Macro. However, they do make the code hard to read. After the implementation is done, I will remove the macro and replace the call with actual code.

For the second part, do you mean I should put the public function declaration at the top and perform the actual implementation afterwards? Also does that mean the definition should take place outside the class declaration? I can certainly do that if it improves readability!

Reply:
I'm going to reorganize the code when I finalize the second part of the feature and before merging the code into the main branch. I will forward this comment to the new PR. Thanks!